compaction/compaction_backlog_manager.hh

```diff
@@ -72,6 +73,9 @@ public:
     compaction_backlog_tracker(const compaction_backlog_tracker&) = delete;
     ~compaction_backlog_tracker();
 
+    // Clone the tracker and return an unregistered, mutable copy of it.
+    compaction_backlog_tracker clone() const;
```

Comment:
Is this really necessary? It looks like the change has huge impact. What does it mean to clone a compaction_backlog_tracker? Will we report double the backlog?

There has to be a better way.

Reply:
Furthermore, what about locking? This is done with preemption points. Can something sneak in and change things under our feet?

Reply:
> What does it mean to clone a compaction_backlog_tracker? Will we report double the backlog?

Since the cloned tracker isn't registered anywhere, we won't double report anything. It's just a bag of data until it's registered.

> Furthermore, what about locking? This is done with preemption points. Can something sneak in and change things under our feet?

Good point. I'm not familiar with the sstables part of this PR, and I assumed during my review that everything here happens under `_sstable_set_mutation_sem`. But it seems this isn't true, and for example `_ongoing_writes` is modified outside of that semaphore, so it could happen concurrently with the read-modify-write of the tracker, dropping info about ongoing writes with some low probability.

Reply:
> What does it mean to clone a compaction_backlog_tracker? Will we report double the backlog?

No, we only clone the contents but not the registration.

> Furthermore, what about locking? This is done with preemption points. Can something sneak in and change things under our feet?

Cache updates are serialized with a semaphore.
This is true for any changes to the compaction group sstable set (adding, removing, or both on compaction completion)

Reply:
> Cache updates are serialized with a semaphore.
This is true for any changes to the compaction group sstable set (adding, removing, or both on compaction completion)

Even for `_ongoing_writes` and `_ongoing_compactions`?

Reply:
> There has to be a better way.

Ulitimately, transactions, as you suggested can do better.  It's still could be a challenge to prevent cloning when there could be asynchronous readers of the state that don't symchronize with the update lock.

For sstable sets, at least, we can carry an auxiliary list of changes that can be in pending (uncommitted) state and so they are ignored by reads and they can be reverted as part of error handling.

Committing them should be noexcept, and once committed they would be considered as part of the set.  Thes set of mutations should be "compacted" into the sstable set main vector.
This can be achieved using a rwlock so cache update can try to acquire the write lock and apply the "mutations" onto the set. If that fails, the mutations are still committed, though not applied.
If the rwlock is taken by a read, cache update just leaves the committed changes behind and any read should detect them, acquire the write lock and apply them then in the beggining of for_each_sstable_gently, or after it's done.

I hope we can intwgrate that mechanism also with the backlog tracker.

Reply:
> > Cache updates are serialized with a semaphore.
> > This is true for any changes to the compaction group sstable set (adding, removing, or both on compaction completion)
> 
> Even for `_ongoing_writes` and `_ongoing_compactions`?

That's an excellent point.
I'll need to rethink this.

Reply:
We have to avoid complexity here. These are rarely-called areas and so we won't have much coverage. I don't want to trade one rare failure mode for another.

Reply:
Backlog tracker reflects the content in sstable set. I suggested before, and I think we should consider marrying sstable_set and backlog_tracker in the compaction namespace. But that's for later.

Reply:
Benny, I don't understand the motivation for this patch, since recently you changed backlog_tracker::replace_sstables to provide strong exception safety. Is this patch really needed? When?

Reply:
Even though it's exception safe it might still throw, and we don't want that during the execute() call.
In retrospect, it's possible we didn't have to make it exception safe if we're doing the update to a temporary value in the prepare phase, but it's a good property to have in any case.